src/utils/synchronization.ts

```diff
@@ -0,0 +1,85 @@
+import axios from "axios";
+import _ from "lodash";
+import "../utils/lodash-mixins";
+
+import Instance from "../models/instance";
+import { D2 } from "../types/d2";
+import { NestedRules, SynchronizationResult } from "../types/synchronization";
+import { cleanModelName, isD2Model } from "./d2";
+import { isValidUid } from "d2/uid";
+
+export function buildNestedRules(rules: string[]): NestedRules {
+    return _.transform(
+        rules,
+        (result, value) => {
+            const [parentType, childType] = value.split(".");
+            result[parentType] = result[parentType] || [];
+            if (childType && !result[parentType].includes(childType)) {
+                result[parentType].push(childType);
+            }
+        },
+        {}
+    );
+}
+
+export async function getMetadata(d2: D2, elements: string[]): Promise<SynchronizationResult> {
+    const promises = [];
+    for (let i = 0; i < elements.length; i += 100) {
+        const requestUrl = d2.Api.getApi().baseUrl + "/metadata.json";
+        const requestElements = elements.slice(i, i + 100).toString();
+        promises.push(
+            axios.get(requestUrl, {
+                withCredentials: true,
+                params: {
+                    fields: ":all",
+                    filter: "id:in:[" + requestElements + "]",
+                    defaults: "EXCLUDE",
+                },
+            })
+        );
+    }
+    const result = await Promise.all(promises);
+    return _.deepMerge({}, ...result.map(result => result.data));
+}
+
+export async function postMetadata(instance: Instance, metadata: any): Promise<any> {
+    return await axios.post(instance.url + "/metadata", metadata, {
```

Comment:
The URL of the instances include the `/api` path? Otherwise it should be added here: `/api/metadata`, right?

Reply:
We were talking how to proceed with @adrianq about this. A new issue has been created to validate that the instance url does not include /api. Will update the code accordingly. See #36